### ISSUE TITLE
Add a reliable way to join telemetry data from different fucntions

### ISSUE THREAD

**Issue by SCasanova** [SCasanova/f1dataR#123]
Speaking to some users this seems to be a big deal and having an in-package function that properly joins data (telemetry with session laps for example) would be very useful.

**Comment by pbulsink** [comment]
Is this an after-CRAN update item, or would you prefer to include this when submitting?

**Comment by SCasanova** [comment]
I'm OK with doing this after CRAN

**Comment by SCasanova** [comment]
I used an interpolation solution in the past but it's not perfect in terms of accuracy. Furthermore it only works with one lap to one lap comparison as it relies on both datasets to have a common variable (in this case distance as the lap should be consitent).

This works comparably to official sources as I proved on my AWS mock app (I think you saw that right @pbulsink ?)

**Comment by pbulsink** [comment]
Yeah I saw it - cool project! I think too I have a better idea of what you're looking for.

FastF1 talks about some of the challenges of time and distance and interpolation in some articles, see: https://docs.fastf1.dev/time_explanation.html and https://docs.fastf1.dev/howto_accurate_calculations.html. 

Of note is the caution against using interpolated data, since the data source frequency is low enough that there's potentially meaningful differences between the interpolated difference something like speed and the actual in rapidly changing environments like when brake is applied at the end of a straight. 

I think if you're doing visual work then overlaid plots show poor enough resolution that interpolation isn't required, but if you're doing modelling or something then only real data should be included. Can you provide an example of your use case where interpolation is beneficial?

**Comment by SCasanova** [comment]
Yeah, I agree with that. The current problem is that measurements aren't taken in the same or even regular intervals so there is no way to compare one to one without interpolating. 

Aside from that, we can look to assign lap number based on distance (for every circuit) to be able to join telemetry to session lap data for example.

**Comment by pbulsink** [comment]
I think there's a way to better interpolate based on series prediction from
`n` points on each side of the target - and doing a left-extrapolation from
data after the time point in question, and right-extrapolation from data
before that time point, and finding where those signals overlap to produce
an interpolated value. It's not going to work for binary data (brake or DRS
on/off) or binary-like data (e.g. 100 to 0 throttle) but might be ok for
RPM, speed, distance (in a time as x axis domain) or time (in a distance as
x axis domain).

Unfortunately, I'm sure that this becomes an active area for mathematics
research and you'll not have high quality data without extensive
computation and estimation (fitting each interpolated point with `n` input
points and fitting `n-1` order polynomials from left and right then finding
where they cross).

For some analysis, an `x = mean(x1,x2)` and `y = mean(y1,y2)` type
interpolation will work, but I agree - for more detailed things this won't.
I think the frequency of the data stream is too low to be making accurate
judgement calls on, for example, who was first on the throttle out of a
certain corner on a given track. When AWS makes those graphics I'm sure
they have higher frequency data to ensure more accurate images that what F1
leaks through their unofficial API.

**Comment by pbulsink** [comment]
To clarify my above comment, this could be the 'worst case scenario' of (for example) speed or RPM being interpolated between a sustained maximum and a sudden decrease (at brake point, or downshift). 

<img width="299" alt="Screenshot 2023-08-24 at 9 57 31 PM" src="https://github.com/SCasanova/f1dataR/assets/5419974/b3180170-2aae-43c5-bbf7-1466490e190c">

**Comment by SCasanova** [comment]
Great illustration, thank you. Perhaps polynomial interpolation would be better in those cases.

I also did a very unexhaustive test of the cross extrapolation you proposed above and... it worked very well. I can share the file

**Comment by pbulsink** [comment]
I noticed a line about interpretation in today's FastF1 release that might help here:

 https://github.com/theOehrly/Fast-F1/releases/tag/v3.5.2

**Comment by theOehrly** [comment]
@pbulsink for some reason I am subscribed to this issue and got a ping. I'm not sure that this fix will help with your problem, though. The change was switching the linear interpolation method that is used for some data channels from 'linear' to 'index'. See the description here: https://pandas.pydata.org/docs/reference/api/pandas.Series.interpolate.html

The index values of those DataFrames are the datetime values of the data. As long as the sample rate is constant, this changes nothing. Only when the sample rate is non-constant, missing values will now be correctly interpolated based on the timestamp.

It's a bit embarrassing that this wrong method was used for multiple years since this feature was introduced.

**Comment by pbulsink** [comment]
Sorry for the ping, I was just linking to the release details for the mention of interpolation. The added background information might help us out, but this is an old issue that's obviously not a high priority for us! Feel free to unfollow.

**Comment by theOehrly** [comment]
No need to apologise. I guess I followed this issue specifically myself some time ago. Don't think it was because of the link. Maybe I wanted to see if you find a good solution to that problem. 
Anyhow, while I was already here, I thought I could give some more details on this change.

**Comment by pbulsink** [comment]
This is a lagging issue in my mind but one user (possibly the one mentioned in the initial issue post) merges based on euclidean distance to the telemetry collected point (see https://github.com/amritvignesh/Formula-One-TSOE/blob/main/turns.R#L62). This doesn't resolve the overshoot of modelling the function from both left and right but provides close enough data to develop a model (unfortunately flat for [TSOE] vs. [season total points] but possibly predictive on a per-race scale). 

Regardless, I'm also becoming more of the mind that this is outside the scope of f1dataR which serves to provide the data to the user. Modelling decisions shouldn't be imposed by the data collection but per the users' needs. I'd suggest to close this issue.
